File: ruby/clock/clock.rb
class Clock
  attr_reader :hours, :minutes

  def initialize(hour: 0, minute: 0)
    @hours = (hour + (minute / 60)) % 24
    @minutes = minute % 60
  end

  def to_s
    pad_hours = hours < 10 ? "0#{hours}" : hours.to_s
    pad_minutes = minutes < 10 ? "0#{minutes}" : minutes.to_s
    "#{pad_hours}:#{pad_minutes}"
  end

  def +(other)
    Clock.new(hour: self.hours + other.hours,
              minute: self.minutes + other.minutes)
  end

  def -(other)
    Clock.new(hour: self.hours - other.hours,
              minute: self.minutes - other.minutes)
  end

  def ==(other)
    self.hours == other.hours &&
      self.minutes == other.minutes
  end
end

File: clojure/clock/src/clock.clj
(ns clock)

(defn clock [hr min]
  (let [time (mod (+ (* hr 60) min) 1440)
        hours (mod (quot time 60) 24)
        minutes (mod time 60)]
    (list hours minutes)))

(defn- padzero [num]
  (if (< num 10) (str 0 num) num))

(defn clock->string [[hours minutes]]
  (str (padzero hours) ":" (padzero minutes)))

(defn add-time [[hour minutes] time]
  (clock hour (+ minutes time)))
